nmdc_schema/migrators/migrator_from_X_to_PR9.py

```diff
@@ -0,0 +1,243 @@
+        r"""
+        Populates the `workflow_chain_set` collection based upon the data in the `self.workflow_omics_dict` dictionary
+        and the `self.omics_analyte_category_dict` dictionary.
+
+        >>> from nmdc_schema.migrators.adapters.dictionary_adapter import DictionaryAdapter
+        >>>
+        >>> database = {"workflow_chain_set": []}  # seeds the database
+        >>> adapter = DictionaryAdapter(database=database)
+        >>> m = Migrator(adapter=adapter)
+        >>> m.workflow_omics_dict = {'nmdc:omcp-123': 'nmdc:wfc-456'}
+        >>> m.omics_analyte_category_dict = {'nmdc:omcp-123': 'metagenome'}
+        >>> m.populate_workflow_chain()
+        >>> adapter.get_document_having_value_in_field('workflow_chain_set', 'id', value='nmdc:wfc-456')
+        {'id': 'nmdc:wfc-456', 'was_informed_by': 'nmdc:omcp-123', 'analyte_category': 'metagenome', 'type': 'nmdc:WorkflowChain', 'name': 'Workflow Chain for metagenome analysis of nmdc:omcp-123'}
+        """
+
+        for omics_id, workflow_chain_id in self.workflow_omics_dict.items():
+            workflow_chain_doc = {}
+
+            if omics_id in self.omics_analyte_category_dict:
+                analyte_category = self.omics_analyte_category_dict.get(omics_id)
+
+            workflow_chain_doc["id"] = workflow_chain_id
+            workflow_chain_doc["was_informed_by"] = omics_id
+
+            workflow_chain_doc["analyte_category"] = analyte_category
+
+            workflow_chain_doc["type"] = "nmdc:WorkflowChain"
+
+            workflow_chain_doc["name"] = (
+                f"Workflow Chain for {analyte_category} analysis of {omics_id}"
+            )
+
+            self.adapter.insert_document("workflow_chain_set", workflow_chain_doc)
+
+    def remove_was_informed_by(self, workflow_doc: dict):
+        r"""
+        Removes the was_informed_by slot from the workflow sets if its value matches its corresponding 
+        workflow_chain_set doc's was_informed_by value.
+
+        >>> from nmdc_schema.migrators.adapters.dictionary_adapter import DictionaryAdapter
+        >>>
+        >>> database = {'workflow_chain_set': [{'id': 'nmdc:wfc-456', 'was_informed_by': 'nmdc:omcp-123'}]}  # in this example, our data store is a Python dictionary
+        >>> adapter = DictionaryAdapter(database=database)
+        >>> m = Migrator(adapter=adapter)
+        >>> m.remove_was_informed_by({'id': 'nmdc:metab-123', 'part_of': ['nmdc:wfc-456'], 'was_informed_by': 'nmdc:omcp-123'})
+        {'id': 'nmdc:metab-123', 'part_of': ['nmdc:wfc-456']}
+        """
+
+        omics_id = workflow_doc["was_informed_by"]
+        workflow_chain_ids = workflow_doc["part_of"]
+        
+        for wfc_id in workflow_chain_ids:
+
+            workflow_chain_doc = self.adapter.get_document_having_value_in_field(
+                collection_name="workflow_chain_set", field_name="id", value=wfc_id
+            )
+            self.logger.info(f"{workflow_chain_doc['id']=}")
```

Comment:
When I ran this migrator on a copy of the production database (a copy that had already undergone migration by the six migrators listed below), this line of code produced a `TypeError`.

![image](https://github.com/microbiomedata/berkeley-schema-fy24/assets/134325062/afb99706-f0b8-4d20-8dd9-68b72f03b233)

I take this to mean `self.adapter.get_document_having_value_in_field()` returned `None`; which would happen if the adapter found no such document in the database. Maybe this is the behavior the author of this migrator expected while the `workflow_chain_ids.yaml` YAML file is empty (in which case, I'll hold off on testing this further).

Here are the six migrators that had already been run on the database in question:
```py
from nmdc_schema.migrators.migrator_from_X_to_PR23 import Migrator as M1
from nmdc_schema.migrators.migrator_from_X_to_PR4 import Migrator as M2
from nmdc_schema.migrators.migrator_from_X_to_PR53 import Migrator as M3
from nmdc_schema.migrators.migrator_from_X_to_PR21 import Migrator as M4
from nmdc_schema.migrators.migrator_from_X_to_PR129 import Migrator as M5
from nmdc_schema.migrators.migrator_from_X_to_PR31 import Migrator as M6
```

Reply:
Oh yes sorry this migrator is broken for now cause I pushed changes to get ahead of the preminting process. I will finish it up today!